notion_client/helpers.py

```diff
@@ -124,3 +134,169 @@ def is_equation_rich_text_item_response(rich_text: Dict[Any, Any]) -> bool:
+    """Extract a Notion ID from a Notion URL or return the input if it's already a valid ID.
+
+    Prioritizes path IDs over query parameters to avoid extracting view IDs instead of database IDs.
+
+    Args:
+        url_or_id: A Notion URL or ID string
+
+    Returns:
+        The extracted UUID in standard format (with hyphens) or None if invalid
+
+    Example:
+        >>> # Database URL with view ID - extracts database ID, not view ID
+        >>> extract_notion_id('https://notion.so/workspace/DB-abc123def456789012345678901234ab?v=viewid123')
+        'abc123de-f456-7890-1234-5678901234ab'  # database ID
+
+        >>> # Already formatted UUID
+        >>> extract_notion_id('12345678-1234-1234-1234-123456789abc')
+        '12345678-1234-1234-1234-123456789abc'
+    """
+    if not url_or_id or not isinstance(url_or_id, str):
+        return None
+
+    trimmed = url_or_id.strip()
+
+    # Check if it's already a properly formatted UUID
+    uuid_pattern = re.compile(
+        r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$", re.IGNORECASE
+    )
+    if uuid_pattern.match(trimmed):
+        return trimmed.lower()
+
+    # Check if it's a compact UUID (32 chars, no hyphens)
+    compact_uuid_pattern = re.compile(r"^[0-9a-f]{32}$", re.IGNORECASE)
+    if compact_uuid_pattern.match(trimmed):
+        return _format_uuid(trimmed.lower())
+
+    # Try to parse as URL
+    try:
+        parsed = urlparse(trimmed)
+
+        # Check if it's a valid Notion domain
+        valid_domains = ["notion.so", "www.notion.so", "notion.site"]
+        if parsed.netloc not in valid_domains:
+            return None
+
+        # First, try to extract ID from the path (prioritized)
+        path = parsed.path
+        if path:
+            # Look for 32-character hex strings in the path
+            path_id_matches = re.findall(r"[0-9a-f]{32}", path, re.IGNORECASE)
+            if path_id_matches:
+                # Take the last match (most likely to be the main ID)
+                return _format_uuid(path_id_matches[-1].lower())
+
+        # If no path ID found, try query parameters
+        query_params = parse_qs(parsed.query)
+        for param in ["p", "id"]:
+            if param in query_params and query_params[param]:
+                candidate = query_params[param][0].strip()
+                if compact_uuid_pattern.match(candidate):
+                    return _format_uuid(candidate.lower())
+                elif uuid_pattern.match(candidate):
+                    return candidate.lower()
+
+        # Try fragment (for block IDs)
+        if parsed.fragment:
+            fragment = parsed.fragment
+            # Remove "block-" prefix if present
+            if fragment.startswith("block-"):
+                fragment = fragment[6:]
+            if compact_uuid_pattern.match(fragment):
+                return _format_uuid(fragment.lower())
+
+        return None
+
+    except Exception:
+        return None
```

Comment:
Why is the implementation different than in notion-sdk-js here?

https://github.com/makenotion/notion-sdk-js/blob/6945289f774aa0150d9c382b1f88c752dbe0f925/src/helpers.ts#L249-L262